### ISSUE TITLE
`clippy::uninit_vec` raises ICE

### ISSUE THREAD

**Issue by usamoi** [rust-lang/rust-clippy#12032]
### Summary

```rust
use std::cell::UnsafeCell;

// #[allow(clippy::uninit_vec)]
unsafe fn _check<S>() {
    let mut vec = Vec::<UnsafeCell<*mut S>>::with_capacity(1);
    vec.set_len(1);
}

fn main() {}
```

This code raises ICE. If `clippy::uninit_vec` is skipped, it does not raise ICE any more.

### Version

```text
rustc 1.77.0-nightly (89e2160c4 2023-12-27)
binary: rustc
commit-hash: 89e2160c4ca5808657ed55392620ed1dbbce78d1
commit-date: 2023-12-27
host: x86_64-unknown-linux-gnu
release: 1.77.0-nightly
LLVM version: 17.0.6
```

### Error output

<details><summary>Backtrace</summary>
  <p>

  ```
    Checking play_rust v0.0.0 (/usamoi/repos/play_rust)
error: internal compiler error: /rustc/89e2160c4ca5808657ed55392620ed1dbbce78d1/compiler/rustc_middle/src/ty/layout.rs:971:21: failed to get layout for `*mut S`: Unknown(Alias(Projection, AliasTy { args: [S/#0], def_id: DefId(2:1880 ~ core[6d34]::ptr::metadata::Pointee::Metadata) })),
                                despite it being a field (#0) of an existing layout: TyAndLayout {
                                    ty: std::cell::UnsafeCell<*mut S>,
                                    layout: Layout {
                                        size: Size(8 bytes),
                                        align: AbiAndPrefAlign {
                                            abi: Align(8 bytes),
                                            pref: Align(8 bytes),
                                        },
                                        abi: Scalar(
                                            Initialized {
                                                value: Pointer(
                                                    AddressSpace(
                                                        0,
                                                    ),
                                                ),
                                                valid_range: 0..=18446744073709551615,
                                            },
                                        ),
                                        fields: Arbitrary {
                                            offsets: [
                                                Size(0 bytes),
                                            ],
                                            memory_index: [
                                                0,
                                            ],
                                        },
                                        largest_niche: None,
                                        variants: Single {
                                            index: 0,
                                        },
                                        max_repr_align: None,
                                        unadjusted_abi_align: Align(8 bytes),
                                    },
                                }

thread 'rustc' panicked at /rustc/89e2160c4ca5808657ed55392620ed1dbbce78d1/compiler/rustc_errors/src/lib.rs:1068:30:
Box<dyn Any>
stack backtrace:
   0: std::panicking::begin_panic::<rustc_errors::ExplicitBug>
   1: <rustc_errors::diagnostic_builder::BugAbort as rustc_errors::diagnostic_builder::EmissionGuarantee>::emit_producing_guarantee
   2: <rustc_errors::DiagCtxt>::bug::<alloc::string::String>
   3: rustc_middle::util::bug::opt_span_bug_fmt::<rustc_span::span_encoding::Span>::{closure#0}
   4: rustc_middle::ty::context::tls::with_opt::<rustc_middle::util::bug::opt_span_bug_fmt<rustc_span::span_encoding::Span>::{closure#0}, !>::{closure#0}
   5: rustc_middle::ty::context::tls::with_context_opt::<rustc_middle::ty::context::tls::with_opt<rustc_middle::util::bug::opt_span_bug_fmt<rustc_span::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
   6: rustc_middle::util::bug::bug_fmt
   7: <rustc_const_eval::interpret::eval_context::InterpCx<rustc_const_eval::const_eval::machine::CompileTimeInterpreter>>::project_field::<rustc_const_eval::interpret::operand::OpTy>
   8: <rustc_const_eval::interpret::validity::ValidityVisitor<rustc_const_eval::const_eval::machine::CompileTimeInterpreter> as rustc_const_eval::interpret::visitor::ValueVisitor<rustc_const_eval::const_eval::machine::CompileTimeInterpreter>>::visit_value
   9: <rustc_const_eval::interpret::eval_context::InterpCx<rustc_const_eval::const_eval::machine::CompileTimeInterpreter>>::validate_operand_internal
  10: rustc_const_eval::util::check_validity_requirement::check_validity_requirement
  11: <rustc_const_eval::provide::{closure#2} as core::ops::function::FnOnce<(rustc_middle::ty::context::TyCtxt, (rustc_middle::ty::layout::ValidityRequirement, rustc_middle::ty::ParamEnvAnd<rustc_middle::ty::Ty>))>>::call_once
      [... omitted 3 frames ...]
  12: clippy_utils::ty::is_uninit_value_valid_for_ty
  13: clippy_lints::uninit_vec::handle_uninit_vec_pair
  14: <clippy_lints::uninit_vec::UninitVec as rustc_lint::passes::LateLintPass>::check_block
  15: <rustc_lint::late::LateContextAndPass<rustc_lint::late::RuntimeCombinedLateLintPass> as rustc_hir::intravisit::Visitor>::visit_block
  16: <rustc_lint::late::LateContextAndPass<rustc_lint::late::RuntimeCombinedLateLintPass> as rustc_hir::intravisit::Visitor>::visit_expr::{closure#0}
  17: <rustc_lint::late::LateContextAndPass<rustc_lint::late::RuntimeCombinedLateLintPass> as rustc_hir::intravisit::Visitor>::visit_expr
  18: <rustc_lint::late::LateContextAndPass<rustc_lint::late::RuntimeCombinedLateLintPass> as rustc_hir::intravisit::Visitor>::visit_nested_body
  19: <rustc_lint::late::LateContextAndPass<rustc_lint::late::RuntimeCombinedLateLintPass> as rustc_hir::intravisit::Visitor>::visit_fn
  20: <rustc_lint::late::LateContextAndPass<rustc_lint::late::RuntimeCombinedLateLintPass> as rustc_hir::intravisit::Visitor>::visit_item
  21: <rustc_lint::late::LateContextAndPass<rustc_lint::late::RuntimeCombinedLateLintPass> as rustc_hir::intravisit::Visitor>::visit_mod
  22: rustc_lint::late::check_crate::{closure#0}
  23: rustc_lint::late::check_crate
  24: rustc_interface::passes::analysis
      [... omitted 1 frame ...]
  25: rustc_interface::interface::run_compiler::<core::result::Result<(), rustc_span::ErrorGuaranteed>, rustc_driver_impl::run_compiler::{closure#0}>::{closure#0}
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

note: we would appreciate a bug report: https://github.com/rust-lang/rust-clippy/issues/new?template=ice.yml

note: please attach the file at `/usamoi/repos/play_rust/rustc-ice-2023-12-28T12_01_19-95315.txt` to your bug report

note: compiler flags: --crate-type bin -C embed-bitcode=no -C debuginfo=2 -C incremental=[REDACTED]

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [check_validity_requirement] checking validity requirement for `core::cell::UnsafeCell<*mut S>`: allows being left uninitialized
#1 [analysis] running analysis passes on this crate
end of query stack
note: Clippy version: clippy 0.1.76 (89e2160 2023-12-27)

error: could not compile `play_rust` (bin "play_rust")
  ```

  </p>
</details>

**Comment by PartiallyTyped** [comment]
So I have triaged the issue to:

https://github.com/rust-lang/rust-clippy/blob/master/clippy_utils/src/ty.rs#L612-L616

More specifically 
```
    cx.tcx
        .check_validity_requirement(_)
```
when given:
```
ParamEnvAnd {
    param_env: ParamEnv {
        caller_bounds: [
            Binder { value: TraitPredicate(<S as std::marker::Sized>, polarity:Positive), bound_vars: [] },
        ],
        reveal: UserFacing,
    },
    value: std::cell::UnsafeCell<*mut S/#0>,
}
```

**Comment by GuillaumeGomez** [comment]
Maybe ask the compiler team directly?

**Comment by dswij** [comment]
https://github.com/rust-lang/rust/issues/119620

**Comment by usamoi** [comment]
The code still raises ICE in the latest nightly clippy. Why close it?
